source/extensions/filters/network/dubbo_proxy/router/router_impl.cc

```diff
@@ -0,0 +1,352 @@
+    // so there is no need to call callbacks_->resetStream() to notify
+    // the upper layer to release the stream.
+    upstream_request_->resetStream();
+    return;
+  }
+
+  if (end_stream) {
+    // Response is incomplete, but no more data is coming.
+    ENVOY_STREAM_LOG(debug, "dubbo router: response underflow", *callbacks_);
+    upstream_request_->onResetStream(
+        Tcp::ConnectionPool::PoolFailureReason::RemoteConnectionFailure);
+    upstream_request_->onResponseComplete();
+    cleanup();
+  }
+}
+
+void Router::onEvent(Network::ConnectionEvent event) {
+  if (!upstream_request_ || upstream_request_->response_complete_) {
+    // Client closed connection after completing response.
+    return;
+  }
+
+  switch (event) {
+  case Network::ConnectionEvent::RemoteClose:
+    upstream_request_->onResetStream(
+        Tcp::ConnectionPool::PoolFailureReason::RemoteConnectionFailure);
+    break;
+  case Network::ConnectionEvent::LocalClose:
+    upstream_request_->onResetStream(
+        Tcp::ConnectionPool::PoolFailureReason::LocalConnectionFailure);
+    break;
+  default:
+    // Connected is consumed by the connection pool.
+    NOT_REACHED_GCOVR_EXCL_LINE;
+  }
+}
+
+const Network::Connection* Router::downstreamConnection() const {
+  return callbacks_ != nullptr ? callbacks_->connection() : nullptr;
+}
+
+void Router::cleanup() {
+  if (upstream_request_) {
+    upstream_request_.reset();
+  }
+}
+
+Router::UpstreamRequest::UpstreamRequest(Router& parent, Tcp::ConnectionPool::Instance& pool,
+                                         MessageMetadataSharedPtr& metadata,
+                                         SerializationType serialization_type,
+                                         ProtocolType protocol_type)
+    : parent_(parent), conn_pool_(pool), metadata_(metadata),
+      deserializer_(
+          NamedDeserializerConfigFactory::getFactory(serialization_type).createDeserializer()),
+      protocol_(NamedProtocolConfigFactory::getFactory(protocol_type).createProtocol()),
+      request_complete_(false), response_started_(false), response_complete_(false) {}
+
+Router::UpstreamRequest::~UpstreamRequest() {}
+
+Network::FilterStatus Router::UpstreamRequest::start() {
+  Tcp::ConnectionPool::Cancellable* handle = conn_pool_.newConnection(*this);
+  if (handle) {
+    // Pause while we wait for a connection.
+    conn_pool_handle_ = handle;
+    return Network::FilterStatus::StopIteration;
+  }
+
+  return Network::FilterStatus::Continue;
+}
+
+void Router::UpstreamRequest::resetStream() {
+  if (conn_pool_handle_) {
+    ASSERT(!conn_data_);
+    conn_pool_handle_->cancel(Tcp::ConnectionPool::CancelPolicy::Default);
+    conn_pool_handle_ = nullptr;
+  }
+
+  if (conn_data_) {
+    ASSERT(!conn_pool_handle_);
+    conn_data_->connection().close(Network::ConnectionCloseType::NoFlush);
+    conn_data_.reset();
+  }
+}
+
+void Router::UpstreamRequest::encodeData(Buffer::Instance& data) {
+  if (!conn_data_) {
```

Comment:
What is the case that conn_data_ will be nullptr?

Reply:
For example, before connection ready, the filter's callback interface transportEnd has been called, thus triggering encodeData function call. At this time, conn_data_ is nullptr.